### PR TITLE
Optimize upload system

### PR DESCRIPTION
**Pull request by gregberge**
This is the description of how the build system work. We call the "client" the uploader, usually the CLI. The build API is separated in two phases:

## 1. Create — POST /builds

The client sends the screenshots keys (sha256) and the commit.
The server creates the build along signed S3 URLs to upload the screenshots.

## *Client requirement*
Client must upload the screenshots using the signed URLs it got from the server before starting the phase 2.

## 2. Update / Finalize — PUT /builds/:buildId

The client sends the `screenshots` (`key` and `name`).
The server verifies that the screenshots are present and mark the build as complete.

## Parallel mode

The parallel mode is used when the client wants to upload the screenshots in several times.
To activate the pallaral mode, the client must send the `parallel` parameter with a value of `true`.
It works with two fondamental parameters:
- `parallelTotal`: the total number of upload batches (usually the number of parallel workers)
- `parallelNonce`: the nonce of the current upload batch

### CODE REVIEW

File: apps/web/src/api/builds.js [hunk 1]
```diff
@@ -48,21 +48,22 @@ async function createBuild({ data, repository, complete = true, trx }) {
...
 async function useExistingBuild({ data, repository, trx }) {
-  const existingBuild = await Build.query(trx)
-    .withGraphFetched("compareScreenshotBucket")
-    .findOne({
-      "builds.repositoryId": repository.id,
-      externalId: data.externalBuildId,
-      name: data.name,
-    });
+  const existingBuild = await Build.query(trx).findOne({
+    "builds.repositoryId": repository.id,
```
Comment (jsfez) [inline apps/web/src/api/builds.js:52]:
```suggestion
    repositoryId: repository.id,
```

File: apps/web/src/api/builds.js [hunk 2]
```diff
@@ -48,21 +48,22 @@ async function createBuild({ data, repository, complete = true, trx }) {
...
+  });
 
   // @TODO Throw an error if batchCount is superior to expected
 
   if (existingBuild) {
     await existingBuild
       .$query(trx)
-      .patchAndFetch({ batchCount: raw('"batchCount" + 1') });
-    return existingBuild;
+      .patch({ batchCount: raw('"batchCount" + 1') });
```
Comment (jsfez) [inline apps/web/src/api/builds.js:62]:
```
return existingBuild
  .$query(trx)
  .patchAndFetch({ batchCount: raw('"batchCount" + 1') })
  .withGraphFetched("compareScreenshotBucket");
```

File: apps/web/src/api/v2/builds/index.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import express from "express";
+import create from "./create";
+import update from "./update";
+
+const router = express.Router();
+export default router;
+
+// This is the description of how the build system work. We call the "client" the uploader,
```
Comment (jsfez) [inline apps/web/src/api/v2/builds/index.js:8]: You should move this explanation into a readme file

**Review by jsfez — APPROVED** [review]

File: apps/storage/src/signed-url.js [hunk 4]
```diff
@@ -1,7 +1,12 @@
 import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
-import { GetObjectCommand } from "@aws-sdk/client-s3";
+import { GetObjectCommand, PutObjectCommand } from "@aws-sdk/client-s3";
 
 export async function getSignedGetObjectUrl({ s3, Bucket, Key, expiresIn }) {
   const command = new GetObjectCommand({ Bucket, Key });
   return getSignedUrl(s3, command, { expiresIn });
 }
+
+export async function getSignedPutObjectUrl({ s3, Bucket, Key, expiresIn }) {
```
[the remark this answers is not in the document]
Comment (gregberge) [reply apps/storage/src/signed-url.js:9]: Oui mais ce n'est pas ce que ça fait. C'est pour obtenir une URL signée qui permet de faire un putObject.

File: apps/web/src/api/builds.js (hunk 2 above)
Comment (gregberge) [reply apps/web/src/api/builds.js:62]: Non ça ne marche pas sûrement un bug d'Objection.

**Pull request merged**